### PR TITLE
Configurable MEM1 and MEM2 sizes at runtime via Dolphin.ini

### PR DESCRIPTION
**Pull request by Minty-Meeo**
Changed several enums from Memmap.h to be static vars and implemented Get functions to query them. This seems to have boosted speed a bit in some titles? The new variables and some previously statically initialized items are now initialized via Memory::Init() and the new AddressSpace::Init(). s_ram_size_real and the new s_exram_size_real in particular are initialized from new OnionConfig values "MAIN_MEM1_SIZE" and "MAIN_MEM2_SIZE", only if "MAIN_RAM_OVERRIDE_ENABLE" is true.

GUI features have been added to Config > Advanced to adjust the new OnionConfig values.

A check has been added to State::doState to ensure savestates with memory configurations different from the current settings aren't loaded. The STATE_VERSION is now 115.

FIFO Files have been updated from version 4 to version 5, now including the MEM1 and MEM2 sizes from the time of DFF creation. FIFO Logs not using the new features (OnionConfig MAIN_RAM_OVERRIDE_ENABLE is false) are still backwards compatible. FIFO Logs that do use the new features have a MIN_LOADER_VERSION of 5. Thanks to the order of function calls, FIFO logs are able to automatically configure the new OnionConfig settings to match what is needed. This is a bit hacky, though, so I also threw in a failsafe for if the conditions that allow this to work ever go away.

I took the liberty of adding a log message to explain why the core fails to initialize if the MIN_LOADER_VERSION is too great.

Some IOS code has had the function "RAMOverrideForIOSMemoryValues" appended to it to recalculate IOS Memory Values from retail IOSes/apploaders to fit the extended memory sizes.  Worry not, if MAIN_RAM_OVERRIDE_ENABLE is false, this function does absolutely nothing.

A hotfix in DolphinQt/MenuBar.cpp has been implemented for RAM Override.

### CODE REVIEW

**Comment by JosJuice** [comment]
@dolphin-emu-bot rebuild

**Comment by AdmiralCurtiss** [comment]
The comments about the static initialization from #8721 still apply here.

**Comment by AdmiralCurtiss** [comment]
This also looks like it would horribly break if you tried to load a savestate with a different ram size than the current configuration.

**Comment by JMC47** [comment]
Savestates need to have something added to protect against memory sizes.

File: Source/Core/Core/HW/Memmap.cpp [hunk 1]
```diff
@@ -57,6 +60,71 @@ u8* m_pL1Cache;
 u8* m_pEXRAM;
 u8* m_pFakeVMEM;
 
+static u32 get_MEM1Size()
+{
+  u32 MEM1_Size;
+  // I need to initialize some stuff that is normally initialized in
+  // int main(), because this funtion will be run before any of that.
+  UICommon::SetUserDirectory("");
```
Comment (lioncash) [inline Source/Core/Core/HW/Memmap.cpp:68]: This is kind of a large code smell. I don't think we should be calling UI specifics from Core.

File: Source/Core/Core/HW/Memmap.cpp [hunk 2]
```diff
@@ -57,6 +60,71 @@ u8* m_pL1Cache;
 u8* m_pEXRAM;
 u8* m_pFakeVMEM;
 
+static u32 get_MEM1Size()
```
Comment (lioncash) [inline Source/Core/Core/HW/Memmap.cpp:63]:
```suggestion
static u32 GetMEM1Size()
```

File: Source/Core/Core/HW/Memmap.cpp [hunk 3]
```diff
@@ -57,6 +60,71 @@ u8* m_pL1Cache;
...
+{
+  u32 MEM1_Size;
```
Comment (lioncash) [inline Source/Core/Core/HW/Memmap.cpp:65]: `mem1_size`, this should also be placed closer to its usage point.

File: Source/Core/Core/HW/Memmap.cpp [hunk 4]
```diff
@@ -57,6 +60,71 @@ u8* m_pL1Cache;
...
+  else
+  {
+    // Retail MEM1 Size
+    MEM1_Size = 0x01800000;
+  }
+
+  SConfig::Shutdown();
+  return MEM1_Size;
+}
+static u32 get_MEM2Size()
```
Comment (lioncash) [inline Source/Core/Core/HW/Memmap.cpp:86]: Same comments apply to this function as the ones for the one above it.

File: Source/Core/Core/HW/AddressSpace.cpp [hunk 5]
```diff
@@ -303,6 +304,7 @@ struct CompositeAddressSpaceAccessors : Accessors
 
 struct SmallBlockAccessors : Accessors
 {
+  SmallBlockAccessors() : alloc_base() {}
```
Comment (lioncash) [inline Source/Core/Core/HW/AddressSpace.cpp:307]: `alloc_base` can just be assigned nullptr directly in the class (along with the size being set to zero). This constructor can then just be defaulted.

File: Source/Core/Core/HW/AddressSpace.cpp [hunk 6]
```diff
@@ -234,6 +234,7 @@ struct AccessorMapping
 
 struct CompositeAddressSpaceAccessors : Accessors
 {
+  CompositeAddressSpaceAccessors() : m_accessor_mappings() {}
```
Comment (lioncash) [inline Source/Core/Core/HW/AddressSpace.cpp:237]:
```suggestion
  CompositeAddressSpaceAccessors() = default;
```

File: Source/Core/Core/HW/Memmap.cpp [hunk 7]
```diff
@@ -189,6 +204,39 @@ static u32 GetFlags()
 
 void Init()
 {
+  auto GetMEM1Size = []()->u32 {
```
Comment (lioncash) [inline Source/Core/Core/HW/Memmap.cpp:207]:
```suggestion
  const auto GetMEM1Size = []() -> u32 {
```

File: Source/Core/Core/HW/Memmap.cpp [hunk 8]
```diff
@@ -189,6 +204,39 @@ static u32 GetFlags()
...
+    if (SConfig::GetInstance().m_RAMOverrideEnable == true)
+      // Configurable MEM1 Size
+      return SConfig::GetInstance().m_MEM1Size;
+    else
+      // Retail MEM1 Size
+      return (u32)0x01800000;
```
Comment (lioncash) [inline Source/Core/Core/HW/Memmap.cpp:213]:
```suggestion
      return 0x01800000U;
```

File: Source/Core/Core/HW/Memmap.cpp [hunk 9]
```diff
@@ -189,6 +204,39 @@ static u32 GetFlags()
 
 void Init()
 {
+  auto GetMEM1Size = []()->u32 {
+    if (SConfig::GetInstance().m_RAMOverrideEnable == true)
+      // Configurable MEM1 Size
```
Comment (lioncash) [inline Source/Core/Core/HW/Memmap.cpp:209]: This if should be braced, since the body spans more than one line.

File: Source/Core/Core/HW/Memmap.cpp [hunk 10]
```diff
@@ -189,6 +204,39 @@ static u32 GetFlags()
...
+      return SConfig::GetInstance().m_MEM1Size;
+    else
```
Comment (lioncash) [inline Source/Core/Core/HW/Memmap.cpp:211]: This else can be removed.

File: Source/Core/Core/HW/Memmap.cpp [hunk 11]
```diff
@@ -189,6 +204,39 @@ static u32 GetFlags()
 
 void Init()
 {
+  auto GetMEM1Size = []()->u32 {
+    if (SConfig::GetInstance().m_RAMOverrideEnable == true)
```
Comment (lioncash) [inline Source/Core/Core/HW/Memmap.cpp:208]:
```suggestion
    if (SConfig::GetInstance().m_RAMOverrideEnable)
```

File: Source/Core/Core/HW/Memmap.cpp [hunk 12]
```diff
@@ -189,6 +204,39 @@ static u32 GetFlags()
...
+      // Configurable MEM1 Size
+      return SConfig::GetInstance().m_MEM1Size;
+    else
+      // Retail MEM1 Size
+      return (u32)0x01800000;
+  };
+  auto GetMEM2Size = []()->u32 {
```
Comment (lioncash) [inline Source/Core/Core/HW/Memmap.cpp:215]: Same review comments for `GetMEM1Size` can be applied here as well.

File: Source/Core/DolphinQt/Settings/AdvancedPane.cpp [hunk 13]
```diff
@@ -208,6 +269,36 @@ void AdvancedPane::Update()
...
+    int MEM1_Size = SConfig::GetInstance().m_MEM1Size / 0x100000;
+    return tr("%1MB (MEM1)").arg(QString::number(MEM1_Size));
+  }());
+
+  m_mem2_override_slider->setEnabled(enable_ram_override_widgets && !running);
+  m_mem2_override_slider_label->setEnabled(enable_ram_override_widgets && !running);
+
+  {
+    const QSignalBlocker blocker(m_mem2_override_slider);
+    int MEM2_Size = SConfig::GetInstance().m_MEM2Size / 0x100000;
```
Comment (lioncash) [inline Source/Core/DolphinQt/Settings/AdvancedPane.cpp:293]:
```suggestion
    const u32 mem2_size = SConfig::GetInstance().m_MEM2Size / 0x100000;
```

File: Source/Core/DolphinQt/Settings/AdvancedPane.cpp [hunk 14]
```diff
@@ -208,6 +269,36 @@ void AdvancedPane::Update()
...
+    m_mem2_override_slider->setValue(MEM2_Size);
+  }
+
+  m_mem2_override_slider_label->setText([] {
+    int MEM2_Size = SConfig::GetInstance().m_MEM2Size / 0x100000;
```
Comment (lioncash) [inline Source/Core/DolphinQt/Settings/AdvancedPane.cpp:298]:
```suggestion
    const u32 mem2_size = SConfig::GetInstance().m_MEM2Size / 0x100000;
```

File: Source/Core/DolphinQt/Settings/AdvancedPane.cpp [hunk 15]
```diff
@@ -208,6 +269,36 @@ void AdvancedPane::Update()
...
+  m_mem1_override_slider_label->setEnabled(enable_ram_override_widgets && !running);
+
+  {
+    const QSignalBlocker blocker(m_mem1_override_slider);
+    int MEM1_Size = SConfig::GetInstance().m_MEM1Size / 0x100000;
+    m_mem1_override_slider->setValue(MEM1_Size);
+  }
+
+  m_mem1_override_slider_label->setText([] {
+    int MEM1_Size = SConfig::GetInstance().m_MEM1Size / 0x100000;
```
Comment (lioncash) [inline Source/Core/DolphinQt/Settings/AdvancedPane.cpp:284]:
```suggestion
    const u32 mem1_size = SConfig::GetInstance().m_MEM1Size / 0x100000;
```

File: Source/Core/DolphinQt/Settings/AdvancedPane.cpp [hunk 16]
```diff
@@ -208,6 +269,36 @@ void AdvancedPane::Update()
...
   }());
 
+  m_ram_override_checkbox->setEnabled(!running);
+
+  m_mem1_override_slider->setEnabled(enable_ram_override_widgets && !running);
+  m_mem1_override_slider_label->setEnabled(enable_ram_override_widgets && !running);
+
+  {
+    const QSignalBlocker blocker(m_mem1_override_slider);
+    int MEM1_Size = SConfig::GetInstance().m_MEM1Size / 0x100000;
```
Comment (lioncash) [inline Source/Core/DolphinQt/Settings/AdvancedPane.cpp:279]:
```suggestion
    const u32 mem1_size = SConfig::GetInstance().m_MEM1Size / 0x100000;
```

File: Source/Core/DolphinQt/Settings/AdvancedPane.cpp [hunk 17]
```diff
@@ -154,6 +193,27 @@ void AdvancedPane::ConnectLayout()
...
+
+  connect(m_mem1_override_slider, &QSlider::valueChanged, [this](int slider_value) {
+    const u32 MEM1_Size = m_mem1_override_slider->value() * 0x100000;
+    SConfig::GetInstance().m_MEM1Size = MEM1_Size;
+    Config::SetBaseOrCurrent(Config::MAIN_MEM1_SIZE, MEM1_Size);
+    Update();
+  });
+
+  connect(m_mem2_override_slider, &QSlider::valueChanged, [this](int slider_value) {
+    const u32 MEM2_Size = m_mem2_override_slider->value() * 0x100000;
```
Comment (lioncash) [inline Source/Core/DolphinQt/Settings/AdvancedPane.cpp:211]:
```suggestion
    const u32 mem2_size = m_mem2_override_slider->value() * 0x100000;
```

File: Source/Core/DolphinQt/Settings/AdvancedPane.cpp [hunk 18]
```diff
@@ -154,6 +193,27 @@ void AdvancedPane::ConnectLayout()
...
 
+  m_ram_override_checkbox->setChecked(SConfig::GetInstance().m_RAMOverrideEnable);
+  connect(m_ram_override_checkbox, &QCheckBox::toggled, [this](bool enable_ram_override) {
+    SConfig::GetInstance().m_RAMOverrideEnable = enable_ram_override;
+    Config::SetBaseOrCurrent(Config::MAIN_RAM_OVERRIDE_ENABLE, enable_ram_override);
+    Update();
+  });
+
+  connect(m_mem1_override_slider, &QSlider::valueChanged, [this](int slider_value) {
+    const u32 MEM1_Size = m_mem1_override_slider->value() * 0x100000;
```
Comment (lioncash) [inline Source/Core/DolphinQt/Settings/AdvancedPane.cpp:204]:
```suggestion
    const u32 mem1_size = m_mem1_override_slider->value() * 0x100000;
```

File: Source/Core/Core/HW/Memmap.cpp (hunk 10 above)
Comment (Minty-Meeo) [reply Source/Core/Core/HW/Memmap.cpp:211]: In my opinion, the else makes the flow of the function more clear at a glance, that is why I left it in.  If you still disagree, I can remove it.

File: Source/Core/Core/HW/Memmap.h [hunk 19]
```diff
@@ -37,23 +37,21 @@ extern u8* m_pEXRAM;
...
-  L1_CACHE_MASK = L1_CACHE_SIZE - 1,
-  IO_SIZE = 0x00010000,
-  EXRAM_SIZE = 0x04000000,
-  EXRAM_MASK = EXRAM_SIZE - 1,
-};
+// g_ram_size is the amount allocated by the emulator, whereas g_ram_size_real is
+// what will be reported in lowmem, and thus used by emulated software.
+// Note: Writing to lowmem is done by IPL. If using retail IPL, it will
+// always be set to 24MB.
+extern u32 g_ram_size_real;
```
Comment (lioncash) [inline Source/Core/Core/HW/Memmap.h:44]: I meant to bring this up in the previous review, but these should likely be hidden from external view and be static file-scope variables and instead have functions that return by value to query the values of these variables. This way other subsystems can query, but not modify these values unintentionally.

File: Source/Core/Core/HW/Memmap.cpp [hunk 20]
```diff
@@ -154,15 +174,10 @@ struct LogicalMemoryView
...
+// TODO: The actual size of RAM is g_ram_size_real (24MB); the other 8MB shouldn't
 // be backed by actual memory.
 // TODO: Do we want to handle the mirrors of the GC RAM?
-static PhysicalMemoryRegion physical_regions[] = {
-    {&m_pRAM, 0x00000000, RAM_SIZE, PhysicalMemoryRegion::ALWAYS},
-    {&m_pL1Cache, 0xE0000000, L1_CACHE_SIZE, PhysicalMemoryRegion::ALWAYS},
-    {&m_pFakeVMEM, 0x7E000000, FAKEVMEM_SIZE, PhysicalMemoryRegion::FAKE_VMEM},
-    {&m_pEXRAM, 0x10000000, EXRAM_SIZE, PhysicalMemoryRegion::WII_ONLY},
-};
+static PhysicalMemoryRegion physical_regions[4];
```
Comment (lioncash) [inline Source/Core/Core/HW/Memmap.cpp:180]:
```suggestion
static std::array<PhysicalMemoryRegion, 4> physical_regions;
```

(`<array>` may need to be included)

Comment (Minty-Meeo) [reply Source/Core/Core/HW/Memmap.cpp:180]: The original implementation did not use std::array.  What benefits does this bring to the table?  Is this just a general improvement that we ought to do while editing Memmap.cpp?

Comment (AdmiralCurtiss) [reply Source/Core/Core/HW/Memmap.cpp:180]: If you're already changing the line, might as well make it better if it's trivial to do so, y'know? Or do you want a general list of reasons why std::array is better than a C-style array?

Comment (Minty-Meeo) [reply Source/Core/Core/HW/Memmap.cpp:180]: Moreso the latter.  I'm still a learning a lot about C++, and have never used std::array before.

Comment (lioncash) [reply Source/Core/Core/HW/Memmap.cpp]: - [std::array](https://en.cppreference.com/w/cpp/container/array) cannot have implicit array to pointer decay, It needs to be explicitly done by using it's `.data()` member function where necessary.

- `std::array` has a `size()` member function, so you can query the size of an array without needing to use `sizeof(array[0]) / sizeof(array)` nonsense.

- It's a wrapper around a C array. This means that implementations can optionally provide bounds checking in debug builds (controllable via a define) to catch out of bounds accesses that may occur across refactoring (currently all three main implementations—MSVC, libstdc++, and libc++—provide this).

TL;DR it's a more strongly-typed and type-safe version of a built-in array.

File: Source/Core/Core/HW/Memmap.h (hunk 19 above)
Comment (lioncash) [reply Source/Core/Core/HW/Memmap.h:44]: This still applies, just to make sure it isn't forgotten about.

Comment (Minty-Meeo) [reply Source/Core/Core/HW/Memmap.h:44]: I'm not a fan of the disconnect between the actual 11 variables and the 11 query functions.  I'd say just don't write bad code, but if this is the last thing keeping you...

Comment (AdmiralCurtiss) [reply Source/Core/Core/HW/Memmap.h:44]: Most of these aren't even read outside of Memmap.cpp as far as I can tell so you don't need to bother exposing them at all.

Comment (lioncash) [reply Source/Core/Core/HW/Memmap.h]: > Most of these aren't even read outside of Memmap.cpp as far as I can tell so you don't need to bother exposing them at all.

This.

Historically, global variables have been a massive pain in Dolphin and a source of frustration for refactoring. If they don't need to be exposed, then they shouldn't be. Especially if they're mutable global variables. It turns a "you shouldn't do that" kind of mistake vector into a "you cannot do that". 

"just don't write bad code", is (IMO) a terrible viewpoint to have especially when the language already provides the necessary facilities to allow one to completely eliminate such code from being written in the first place directly at the interface boundary.

File: Source/Core/Core/BootManager.cpp [hunk 21]
```diff
@@ -95,6 +95,9 @@ struct ConfigCache
   float m_EmulationSpeed;
   float m_OCFactor;
   bool m_OCEnable;
+  bool m_RAMOverrideEnable;
```
Comment (leoetlino) [inline Source/Core/Core/BootManager.cpp:98]: This isn't necessary if you use the new config system (the Common/Config and Core/Config stuff). Ditto for all the other additions to BootManager

File: Source/Core/Core/ConfigManager.cpp [hunk 22]
```diff
@@ -242,6 +242,9 @@ void SConfig::SaveCoreSettings(IniFile& ini)
   core->Set("EmulationSpeed", m_EmulationSpeed);
   core->Set("Overclock", m_OCFactor);
   core->Set("OverclockEnable", m_OCEnable);
+  core->Set("RAMOverrideEnable", m_RAMOverrideEnable);
```
Comment (leoetlino) [inline Source/Core/Core/ConfigManager.cpp:245]: ditto for ConfigManager

File: Source/Core/Core/HW/Memmap.cpp [hunk 23]
```diff
@@ -57,6 +58,69 @@ u8* m_pL1Cache;
...
+// is what gets used by emulated software.  This is usually 64MB.
+static u32 s_exram_size_real;
+static u32 s_exram_size;
+static u32 s_exram_mask;
+
+u32 GetRamSizeReal()
+{
+  return s_ram_size_real;
+}
+u32 GetRamSize()
```
Comment (leoetlino) [inline Source/Core/Core/HW/Memmap.cpp:83]: I'm surprised clang-format is happy with the missing line breaks between these functions... Oh well.

File: Source/Core/Core/HW/Memmap.cpp [hunk 24]
```diff
@@ -189,6 +248,34 @@ static u32 GetFlags()
 
 void Init()
 {
+  const auto GetMEM1Size = []() -> u32 {
```
Comment (leoetlino) [inline Source/Core/Core/HW/Memmap.cpp:251]: Our naming convention is to use snake_case for variables, even if they're lambda functions.

```suggestion
  const auto get_mem1_size = []() -> u32 {
```

(also I'm not sure explicitly specifying the return type is necessary?)

File: Source/Core/Core/HW/Memmap.cpp [hunk 25]
```diff
@@ -189,6 +248,34 @@ static u32 GetFlags()
...
+    if (SConfig::GetInstance().m_RAMOverrideEnable)
+      return SConfig::GetInstance().m_MEM1Size;
+    return 0x01800000U;
+  };
+  const auto GetMEM2Size = []() -> u32 {
```
Comment (leoetlino) [inline Source/Core/Core/HW/Memmap.cpp:256]:
```suggestion
  const auto get_mem2_size = []() -> u32 {
```

File: Source/Core/Core/HW/Memmap.cpp (hunk 23 above)
Comment (Minty-Meeo) [reply Source/Core/Core/HW/Memmap.cpp:83]: I'm surprised clang-format is happy turning what was beautifully described in 11 lines into a mess that takes 44.

File: Source/Core/Core/HW/Memmap.cpp (hunk 24 above)
Comment (Minty-Meeo) [reply Source/Core/Core/HW/Memmap.cpp:251]: If lambdas are to use snake_case, this should be stated in the Contributing.md document.

File: Source/Core/Core/BootManager.cpp (hunk 21 above)
Comment (Minty-Meeo) [reply Source/Core/Core/BootManager.cpp:98]: I don't understand what you're saying isn't necessary here.  I don't even know where the new config system is.

**Comment by phire** [comment]
Like savestatese, Fifologs should also have protection against playing when the memory size is too small.

File: Source/Core/Core/HW/HW.cpp [hunk 26]
```diff
@@ -42,6 +43,7 @@ void Init()
   ProcessorInterface::Init();
   ExpansionInterface::Init();  // Needs to be initialized before Memory
   Memory::Init();
+  AddressSpace::Init();
```
Comment (phire) [inline Source/Core/Core/HW/HW.cpp:46]: This initializer depends on `Memory::Init()`

Ideally we should have a comment whenever there are initialization order dependencies.

File: Source/Core/Core/HW/Memmap.cpp (hunk 23 above)
Comment (lioncash) [reply Source/Core/Core/HW/Memmap.cpp:83]: Completely unnecessary response. We have code-base wide formatting to be consistent across the board and not have individual files with bespoke formatting choices that differ based on what programmer decided to code the file at that given point in time. Please direct snark like this elsewhere in the future, because it's not useful or helpful in a review.

File: Source/Core/Core/BootManager.cpp (hunk 21 above)
Comment (JosJuice) [reply Source/Core/Core/BootManager.cpp:98]: The new config system is in Common/Config and Core/Config, as leoetlino mentioned. If you want to know how to add a new setting to it, you can use e.g. 3c9eb37 as a reference. (You only need to look at the changes in Core/Config.)

Comment (Minty-Meeo) [reply Source/Core/Core/BootManager.cpp:98]: Ooooooooh that's a folder you were referencing.  I was confused.  Alright, I'll look into this.

Comment (Minty-Meeo) [reply Source/Core/Core/BootManager.cpp:98]: Quick question, does the new config system have a cache to restore the "trampled on values" like the old system has?  I'm taking advantage of that in the fix for the FIFO player.

Comment (JosJuice) [reply Source/Core/Core/BootManager.cpp:98]: Actually, it has something better: Layers. At boot, instead of overwriting the values in the base layer, the value changes are added to a different layer. The new config system takes both layers into account when you want to fetch the value of a setting (or not – it is possible to e.g. fetch the value of the base layer specifically). Then when emulation ends, the layer that was used for overriding settings for this emulation session is discarded, making those overridden values go back to normal.

**Comment by JMC47** [comment]
When I adjust MEM1 size, Some Wii Games aren't booting.  MEM2 didn't seem to cause problems.

Tested on Fire Emblem: Radiant Dawn.

**Comment by Minty-Meeo** [comment]
I don't know what causes that, but it seems to be in all Wii games.  Were the Wii Dev units able to use more than 24MB of MEM1?

**Comment by JMC47** [comment]
I think super smash bros. Brawl will print out the memarena in the logs.  I don't know if that's just MEM1, but maybe it'd give us a clue if it does include mem2

**Comment by JMC47** [comment]
Now things appear to be working.  Obviously I can't test MEM2, but with MEM1, the few cases I know where extra memory affects games, they behaved as expected.

File: Source/Core/Core/IOS/VersionInfo.cpp [hunk 27]
```diff
@@ -9,334 +9,339 @@
...
-         0x935E0000,
-         0x935E0000,
-         0x93600000,
-         HOLLYWOOD_REVISION,
-         RAM_VENDOR_MIOS,
-         PLACEHOLDER,
-         PLACEHOLDER,
-         PLACEHOLDER,
-     }}};
+std::array<MemoryValues, 41> ios_memory_values;
```
Comment (leoetlino) [inline Source/Core/Core/IOS/VersionInfo.cpp:17]: Maybe consider updating SetupMemory (in IOS) instead? These values are supposed to be dumped directly from IOS kernels.

**Comment by JMC47** [comment]
@lioncash Are all of your qualms with this taken care of?

File: Source/Core/Core/IOS/IOS.cpp [hunk 28]
```diff
@@ -158,18 +155,77 @@ static bool SetupMemory(u64 ios_title_id, MemorySetupType setup_type)
...
   Memory::Write_U8(0xAD, ADDR_APPLOADER_FLAG);
   Memory::Write_U16(0xBEEF, ADDR_DEVKIT_BOOT_PROGRAM_VERSION);
   Memory::Write_U32(target_imv->sysmenu_sync, ADDR_SYSMENU_SYNC);
+
+  RamOverrideForIOSMemoryValues(setup_type);
+
   return true;
 }
 
+// We're gonna make our own IOS Memory Values, with blackjack and hookers!
```
Comment (AdmiralCurtiss) [inline Source/Core/Core/IOS/IOS.cpp:173]: Please don't write comments like that.

Comment (Minty-Meeo) [reply Source/Core/Core/IOS/IOS.cpp:173]: ok

File: Source/Core/Core/IOS/IOS.cpp [hunk 29]
```diff
@@ -158,18 +155,77 @@ static bool SetupMemory(u64 ios_title_id, MemorySetupType setup_type)
...
+    u32 ios_reserved_size;
+
+    {
+      const u32 ipc_buffer_begin = Memory::Read_U32(ADDR_IPC_BUFFER_BEGIN);
+      const u32 ipc_buffer_end = Memory::Read_U32(ADDR_IPC_BUFFER_END);
+      const u32 ios_reserved_begin = Memory::Read_U32(ADDR_IOS_RESERVED_BEGIN);
+      const u32 ios_reserved_end = Memory::Read_U32(ADDR_IOS_RESERVED_END);
+
+      // Inferring them
+      ipc_buffer_size = ipc_buffer_end - ipc_buffer_begin;
```
Comment (leoetlino) [inline Source/Core/Core/IOS/IOS.cpp:189]: Why not just:

```cpp
const u32 ipc_buffer_size = ipc_buffer_end - ipc_buffer_begin;
```

or even 

```cpp
const u32 ipc_buffer_size =
  Memory::Read_U32(ADDR_IPC_BUFFER_END) - Memory::Read_U32(ADDR_IPC_BUFFER_BEGIN);
```

I don't think it's worth introducing a new scope *and* making some variables non-const.

File: Source/Core/Core/IOS/IOS.cpp [hunk 30]
```diff
@@ -158,18 +155,77 @@ static bool SetupMemory(u64 ios_title_id, MemorySetupType setup_type)
...
   Memory::Write_U8(0xAD, ADDR_APPLOADER_FLAG);
   Memory::Write_U16(0xBEEF, ADDR_DEVKIT_BOOT_PROGRAM_VERSION);
   Memory::Write_U32(target_imv->sysmenu_sync, ADDR_SYSMENU_SYNC);
+
+  RamOverrideForIOSMemoryValues(setup_type);
+
   return true;
 }
 
+// We're gonna make our own IOS Memory Values.
```
Comment (leoetlino) [inline Source/Core/Core/IOS/IOS.cpp:173]: I don't think this comment is necessary or adds much information that is not already in the function name.

File: Source/Core/DolphinQt/Settings/AdvancedPane.cpp [hunk 31]
```diff
@@ -96,6 +96,44 @@ void AdvancedPane::CreateLayout()
   cpu_clock_override_description->setWordWrap(true);
   clock_override_layout->addWidget(cpu_clock_override_description);
 
+  auto* ram_override = new QGroupBox(tr("Memory Override"));
+  auto* ram_override_layout = new QVBoxLayout();
+  ram_override->setLayout(ram_override_layout);
+  main_layout->addWidget(ram_override);
+
+  m_ram_override_checkbox = new QCheckBox(tr("Enable Emulated Memory Override"));
```
Comment (leoetlino) [inline Source/Core/DolphinQt/Settings/AdvancedPane.cpp:104]: "Override Emulated Memory Size" would be clearer, I think.

File: Source/Core/Core/FifoPlayer/FifoDataFile.cpp [hunk 32]
```diff
@@ -199,20 +216,56 @@ std::unique_ptr<FifoDataFile> FifoDataFile::Load(const std::string& filename, bo
...
     file.Close();
     return dataFile;
   }
 
+  // To make up for such a hacky thing, here is a catch-all failsafe in case if the above code
+  // stops working or is otherwise removed.  As it is, this should never end up running.
+  if (header.mem1_size != Memory::GetRamSizeReal() ||
+      header.mem2_size != Memory::GetExRamSizeReal())
+  {
+    ERROR_LOG(BOOT,
```
Comment (JosJuice) [inline Source/Core/Core/FifoPlayer/FifoDataFile.cpp:257]: Due to how serious this is if it happens, I would suggest making this use `PanicAlert` instead of `ERROR_LOG`.

File: Source/Core/Core/IOS/IOS.cpp [hunk 33]
```diff
@@ -158,18 +155,68 @@ static bool SetupMemory(u64 ios_title_id, MemorySetupType setup_type)
...
   Memory::Write_U8(0xAD, ADDR_APPLOADER_FLAG);
   Memory::Write_U16(0xBEEF, ADDR_DEVKIT_BOOT_PROGRAM_VERSION);
   Memory::Write_U32(target_imv->sysmenu_sync, ADDR_SYSMENU_SYNC);
+
+  RamOverrideForIOSMemoryValues(setup_type);
+
   return true;
 }
 
+void RamOverrideForIOSMemoryValues(MemorySetupType setup_type)
```
Comment (JosJuice) [inline Source/Core/Core/IOS/IOS.cpp:173]: For consistency (`IOS` is capitalized), `RAM` should be capitalized.

Comment (leoetlino) [reply Source/Core/Core/IOS/IOS.cpp:173]: Unrelated, but we should probably get rid of the capitalization inconsistencies at some point.

File: Source/Core/Core/FifoPlayer/FifoDataFile.cpp (hunk 32 above)
Comment (Minty-Meeo) [reply Source/Core/Core/FifoPlayer/FifoDataFile.cpp:257]: Should I similarly change the alert I added for if the min_loader_version is too great?

Comment (JosJuice) [reply Source/Core/Core/FifoPlayer/FifoDataFile.cpp]: I suppose that makes sense, yes.

Comment (leoetlino) [reply Source/Core/Core/FifoPlayer/FifoDataFile.cpp]: Maybe even CriticalAlertT (T to make it translatable), so that this is always visible even if the user has disabled panic alerts?

**Comment by Minty-Meeo** [comment]
About the conflicting file.  The only "conflict" it sees is an unused #include I removed.  I'm not sure why it is just now finding issue with that, it never did before.

**Comment by JosJuice** [comment]
It's because that line of code was changed in master 4 hours ago. You'll need to rebase on latest master and fix the conflict.

**Comment by RinMaru** [comment]
Messing with these settings has caused some issues. Tak and the power of Juju will crash mid game with Mem1/2 options modified so i dont think it should be on by default

**Comment by JosJuice** [comment]
@RinMaru It isn't on by default, and turning it on for random games isn't expected to work.
